clarity/contracts/arkadiko-stacker-payer-v2-1.clar

```diff
@@ -0,0 +1,74 @@
+;; Implementation of Stacker Payer
+;; which allows users to redeem xSTX for STX
+
+(define-constant ERR-NOT-AUTHORIZED u22401)
+(define-constant ERR-EMERGENCY-SHUTDOWN-ACTIVATED u221)
+
+(define-data-var stacker-payer-shutdown-activated bool false)
+(define-data-var stx-redeemable uint u0)
+
+(define-read-only (get-stx-redeemable)
+  (var-get stx-redeemable)
+)
+
+(define-read-only (has-stx-redeemable)
+  (> (var-get stx-redeemable) u0)
+)
+
+(define-read-only (is-enabled)
+  (and
+    (not (unwrap-panic (contract-call? .arkadiko-dao get-emergency-shutdown-activated)))
+    (not (var-get stacker-payer-shutdown-activated))
+  )
+)
+
+(define-public (toggle-stacker-payer-shutdown)
+  (begin
+    (asserts! (is-eq tx-sender (contract-call? .arkadiko-dao get-guardian-address)) (err ERR-NOT-AUTHORIZED))
+
+    (ok (var-set stacker-payer-shutdown-activated (not (var-get stacker-payer-shutdown-activated))))
+  )
+)
+
+(define-public (return-stx-to-reserve (ustx-amount uint))
+  (begin
+    (asserts! (is-enabled) (err ERR-EMERGENCY-SHUTDOWN-ACTIVATED))
+
+    (if (> ustx-amount u0)
+      (as-contract
+        (stx-transfer? ustx-amount tx-sender (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "stx-reserve")))
+      )
+      (ok true)
+    )
+  )
+)
+
+(define-public (set-stx-redeemable (ustx-amount uint))
```

Comment:
Suggestion: wondering if we can decentralise this centralised part. The contract could have a public method which takes the vault-id as parameter and calculates the the difference between total collateral and stacked tokens. The calculated difference is added to stx-redeemable and the method can only be called once per vault.

Reply:
@nieldeckx that's fair, but requires a lot of on-chain transactions now to recover? We could initialize it to a correct value perhaps. I'll look into it